include/nbl/ext/Scan/Scan.h

```diff
@@ -0,0 +1,190 @@
+	}
+
+	static inline void postPassParameterUpdate(uint32_t pass_idx, bool is_upsweep, Parameters_t* push_constants, DispatchInfo_t* dispatch_info,
+		const uint32_t wg_size)
+	{
+		if (is_upsweep && (pass_idx != (dispatch_info->upsweep_pass_count - 1u)))
+		{
+			push_constants->stride *= wg_size;
+			push_constants->element_count_pass = dispatch_info->wg_count;
+		}
+		else
+		{
+			push_constants->stride /= wg_size;
+		}
+	}
+
+	static inline void defaultBarrier()
+	{
+		video::COpenGLExtensionHandler::extGlMemoryBarrier(GL_SHADER_STORAGE_BARRIER_BIT);
+	}
+
+	static inline void updateDescriptorSet(video::IGPUDescriptorSet* set, core::smart_refctd_ptr<video::IGPUBuffer> descriptor, video::IVideoDriver* driver)
+	{
+		video::IGPUDescriptorSet::SDescriptorInfo ds_info = {};
+		ds_info.desc = descriptor;
+		ds_info.buffer = { 0u, descriptor->getSize() };
+
+		video::IGPUDescriptorSet::SWriteDescriptorSet writes = { set, 0, 0u, 1u, asset::EDT_STORAGE_BUFFER, &ds_info };
+
+		driver->updateDescriptorSets(1, &writes, 0u, nullptr);
+	}
+
+	const uint32_t m_wg_size;
+
+private:
+	~Scan() {}
+
+	core::smart_refctd_ptr<video::IGPUPipelineLayout> m_pipeline_layout = nullptr;
+	core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> m_ds_layout = nullptr;
+	core::smart_refctd_ptr<video::IGPUComputePipeline> m_upsweep_pipeline = nullptr;
+	core::smart_refctd_ptr<video::IGPUComputePipeline> m_downsweep_pipeline = nullptr;
+
+	core::smart_refctd_ptr<video::IGPUComputePipeline> createPipeline(const char* shader_include_name, Operator bin_op, const char* data_type_name,
+		video::IDriver* driver)
+	{
+		const char* source_fmt =
+R"===(#version 430 core
+
+#define _NBL_GLSL_WORKGROUP_SIZE_ %u
+#define _NBL_GLSL_EXT_SCAN_BIN_OP_ %u
+#define _NBL_GLSL_EXT_SCAN_STORAGE_TYPE_ %s
+
+layout (local_size_x = _NBL_GLSL_WORKGROUP_SIZE_) in;
+ 
+#include "%s"
+
+)===";
+
+		// Todo: This just the value I took from FFT example, don't know how it is being computed.
+		const size_t extraSize = 4u + 8u + 8u + 128u;
+
+		auto shader = core::make_smart_refctd_ptr<asset::ICPUBuffer>(strlen(source_fmt) + extraSize + 1u);
+		snprintf(reinterpret_cast<char*>(shader->getPointer()), shader->getSize(), source_fmt, m_wg_size, bin_op, data_type_name, shader_include_name);
+
+		auto cpu_specialized_shader = core::make_smart_refctd_ptr<asset::ICPUSpecializedShader>(
+			core::make_smart_refctd_ptr<asset::ICPUShader>(std::move(shader), asset::ICPUShader::buffer_contains_glsl),
+			asset::ISpecializedShader::SInfo{ nullptr, nullptr, "main", asset::ISpecializedShader::ESS_COMPUTE });
+
+		auto gpu_shader = driver->createGPUShader(core::smart_refctd_ptr<const asset::ICPUShader>(cpu_specialized_shader->getUnspecialized()));
+		auto gpu_shader_specialized = driver->createGPUSpecializedShader(gpu_shader.get(), cpu_specialized_shader->getSpecializationInfo());
+
+		return driver->createGPUComputePipeline(nullptr, core::smart_refctd_ptr(m_pipeline_layout), std::move(gpu_shader_specialized));
+	}
```

Comment:
This one is for @sadiuk

We seem to have a need for a lot of this in the engine and examples... maybe you could make a `sprintf`-like static function for IGLSLCompiler with the signature
```cpp
template<typename... Args>
inline core::smart_refctd_ptr<ICPUShader > createOverridenCopy(const ICPUShader* original, const char* fmt, Args&&... args)
```

What this function would do is:
1. require that the `original` has a buffer that is not a dummy object (`!original->isDummy()`) and that it contains GLSL
2. Compute the length of the original shader + length of the format string + max possible length that arguments could add when parsed from the format string, then allocate a new `ICPUBuffer` of this "safe size"
3. find the position just after the newline after the `#version` directive, if not found then set to 0
4. Copy the original shader into the new buffer but only as far as the end of the `#version` directive (the position found in step 3)
5. Use
```cpp
sprintf(reinterpret_cast<char*>(newBuffer->getPointer())+pJustAfterNewline,std::string(fmt)+"%s",std::forward<Args>(args)...,reinterpret_cast<char*>(oldBuffer->getPointer())+pJustAfterNewline);
```
to "insert" the new stuff into the shader

Reply:
@sadiuk if you took note, please "Resolve" !

Reply:
@achalpandeyy yeah i have alredy implemented it here: https://github.com/Devsh-Graphics-Programming/Nabla/pull/68/commits/c8e41fafdeda0e837c8015560c299b0b150efdde#diff-a99ed313ad44fc6070ae32c1711bbda52c8669c2e0396dfc8a362cc8405bed1fR88

It is still in my pr, so You should probably wait a bit until it's merged. Sorry, didn't know it was so urgent.

Reply:
nope wasn't urgent I just wanted to make sure you have seen this comment